File: src/main/java/com/raoulvdberge/refinedstorage/item/blockitem/EnergyBlockItem.java
package com.raoulvdberge.refinedstorage.item.blockitem;

import com.raoulvdberge.refinedstorage.item.capabilityprovider.EnergyCapabilityProvider;
import net.minecraft.block.Block;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.item.BlockItem;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundNBT;
import net.minecraft.util.math.MathHelper;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.Style;
import net.minecraft.util.text.TextFormatting;
import net.minecraft.util.text.TranslationTextComponent;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.ICapabilityProvider;
import net.minecraftforge.energy.CapabilityEnergy;
import net.minecraftforge.energy.IEnergyStorage;

import javax.annotation.Nullable;
import java.util.List;
import java.util.function.Supplier;

public abstract class EnergyBlockItem extends BlockItem {
    private final Supplier<Integer> energyCapacity;
    private final boolean creative;

    public EnergyBlockItem(Block block, Item.Properties properties, boolean creative, Supplier<Integer> energyCapacity) {
        super(block, properties);

        this.creative = creative;
        this.energyCapacity = energyCapacity;
    }

    @Override
    public ICapabilityProvider initCapabilities(ItemStack stack, CompoundNBT tag) {
        return new EnergyCapabilityProvider(stack, energyCapacity.get());
    }

    @Override
    public boolean showDurabilityBar(ItemStack stack) {
        return !creative;
    }

    @Override
    public double getDurabilityForDisplay(ItemStack stack) {
        IEnergyStorage energy = stack.getCapability(CapabilityEnergy.ENERGY).orElse(null);
        if (energy == null) {
            return 0;
        }

        return 1D - ((double) energy.getEnergyStored() / (double) energy.getMaxEnergyStored());
    }

    @Override
    public int getRGBDurabilityForDisplay(ItemStack stack) {
        IEnergyStorage energy = stack.getCapability(CapabilityEnergy.ENERGY, null).orElse(null);
        if (energy == null) {
            return super.getRGBDurabilityForDisplay(stack);
        }

        return MathHelper.hsvToRGB(Math.max(0.0F, (float) energy.getEnergyStored() / (float) energy.getMaxEnergyStored()) / 3.0F, 1.0F, 1.0F);
    }

    @Override
    public void addInformation(ItemStack stack, @Nullable World world, List<ITextComponent> tooltip, ITooltipFlag flag) {
        super.addInformation(stack, world, tooltip, flag);

        if (!creative) {
            stack.getCapability(CapabilityEnergy.ENERGY, null).ifPresent(energy -> tooltip.add(new TranslationTextComponent("misc.refinedstorage.energy_stored", energy.getEnergyStored(), energy.getMaxEnergyStored()).setStyle(new Style().setColor(TextFormatting.GRAY))));
        }
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/item/blockitem/EnergyBlockItem.kt
package com.refinedmods.refinedstorage.item.blockitem

import com.refinedmods.refinedstorage.block.BaseBlock
import com.refinedmods.refinedstorage.item.capabilityprovider.EnergyCapabilityProvider
import com.refinedmods.refinedstorage.render.Styles
import net.minecraft.client.item.TooltipContext
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.text.Text
import net.minecraft.util.math.MathHelper
import net.minecraft.world.World
import reborncore.client.gui.builder.widget.tooltip.ToolTip
import java.util.function.Supplier

abstract class EnergyBlockItem(block: BaseBlock, properties: Item.Settings, private val creative: Boolean, private val energyCapacity: Supplier<Int>):
        BaseBlockItem(block, properties)
{
    // TODO Replace capability
//    fun initCapabilities(stack: ItemStack?, tag: CompoundTag?): ICapabilityProvider {
//        return EnergyCapabilityProvider(stack, energyCapacity.get())
//    }

    fun showDurabilityBar(stack: ItemStack): Boolean {
        return !creative
    }

    fun getDurabilityForDisplay(stack: ItemStack): Double {
        // TODO energy
//        val energy: IEnergyStorage = stack.getCapability(CapabilityEnergy.ENERGY).orElse(null) ?: return 0
//        return 1.0 - energy.getEnergyStored() as Double / energy.getMaxEnergyStored() as Double

        return 0.0
    }


    fun getRGBDurabilityForDisplay(stack: ItemStack): Int {
        // TODO energy
//        val energy: IEnergyStorage = stack.getCapability(CapabilityEnergy.ENERGY, null).orElse(null)
//                ?: return super.getRGBDurabilityForDisplay(stack)
//        return MathHelper.hsvToRGB(Math.max(0.0f, energy.getEnergyStored() as Float / energy.getMaxEnergyStored() as Float) / 3.0f, 1.0f, 1.0f)
        return 0
    }

    override fun appendTooltip(stack: ItemStack?, world: World?, tooltip: MutableList<Text>?, context: TooltipContext?) {
        super.appendTooltip(stack, world, tooltip, context)
        if (!creative) {
            // TODO Energy
//            stack.getCapability(CapabilityEnergy.ENERGY, null).ifPresent({ energy -> tooltip.add(TranslationTextComponent("misc.refinedstorage.energy_stored", energy.getEnergyStored(), energy.getMaxEnergyStored()).setStyle(Styles.GRAY)) })
        }
    }

}